python/ray/util/spark/cluster_init.py

```diff
@@ -1044,23 +1075,73 @@ def _setup_ray_cluster_internal(
                 "scheduling, you need to upgrade spark to 3.4 version or use "
                 "Databricks Runtime 12.x, and you cannot use spark local mode."
             )
-    else:
-        using_stage_scheduling = False
-        res_profile = None
+    elif num_cpus_worker_node is None and num_gpus_worker_node is None:
+        if support_stage_scheduling:
+            # Make one Ray worker node using maximum CPU / GPU resources
+            # of the whole spark worker node, this is the optimal
+            # configuration.
+            num_cpus_worker_node = num_cpus_spark_worker
+            num_gpus_worker_node = num_gpus_spark_worker
+            using_stage_scheduling = True
+            res_profile = _create_resource_profile(
+                num_cpus_worker_node, num_gpus_worker_node
+            )
+        else:
+            using_stage_scheduling = False
+            res_profile = None
 
-        num_cpus_worker_node = num_spark_task_cpus
-        num_gpus_worker_node = num_spark_task_gpus
+            num_cpus_worker_node = num_spark_task_cpus
+            num_gpus_worker_node = num_spark_task_gpus
+    else:
+        raise ValueError(
+            "'num_cpus_worker_node' and 'num_gpus_worker_node' arguments must be"
+            "set together or unset together."
```

Comment:
Why is that?

Reply:
The reason is:

in this PR I change `num_cpus_worker_node` and `num_gpus_worker_node` default values to "maximum" CPU  / GPU cores of spark worker node:

Assuming the spark worker node instance has 4 CPUs and 2 GPU,
then if user only set one param like:

`setup_ray_cluster( =1)`

then `num_gpus_worker_node` default values still use 2, then it causes 3 CPU cores in this node wasted, because once the shape (1 CPUs/ 2 GPUs) ray node launched in this spark worker, no other ray worker node of the same shape can be launched in the same spark worker.


So I only allow 2 cases:

1. num_cpus_worker_node / num_gpus_worker_node are generated by default, to use all CPU /GPU cores in the spark worker.
2. Let user decide both of num_cpus_worker_node and num_gpus_worker_node, then user will set proper value e.g. (num_cpus_worker_node=2, num_gpus_worker_node=1) which can fit 4 CPU cores 2 GPU cores spark worker shape well.

Reply:
I feel it's better to allow users to set them separately since they are two independent resources. If you want to avoid wasted resources, we can print some warning message. (What if the waste is on purpose that user wants to leave some resources for spark jobs)?